integrations/pinecone/src/pinecone_haystack/document_store.py

```diff
@@ -0,0 +1,236 @@
+        :param batch_size: The number of documents to write in a single batch. Defaults to 100, as recommended by
+            Pinecone.
+        :param dimension: The dimension of the embeddings. This parameter is only used when creating a new index.
+            Defaults to 768.
+        :param index_creation_kwargs: Additional keyword arguments to pass to the index creation method.
+            For example, you can specify `metric`, `pods`, `replicas`...
+            You can find the full list of supported arguments in the
+            [API reference](https://docs.pinecone.io/reference/create_index-1).
+
+        """
+        api_key = api_key or os.environ.get("PINECONE_API_KEY", None)
+        if not api_key:
+            msg = (
+                "PineconeDocumentStore expects a Pinecone API key. "
+                "Set the PINECONE_API_KEY environment variable (recommended) or pass it explicitly."
+            )
+            raise ValueError(msg)
+
+        pinecone.init(api_key=api_key, environment=environment)
+
+        if index not in pinecone.list_indexes():
+            pinecone.create_index(name=index, dimension=dimension, **index_creation_kwargs)
+
+        self._index = pinecone.Index(index_name=index)
+        self.dimension = self._index.describe_index_stats()["dimension"]
+        self._dummy_vector = [0.0] * self.dimension
+        self.environment = environment
+        self.index = index
+        self.namespace = namespace
+        self.batch_size = batch_size
+        self.index_creation_kwargs = index_creation_kwargs
+
+    def to_dict(self) -> Dict[str, Any]:
+        return default_to_dict(
+            self,
+            environment=self.environment,
+            index=self.index,
+            dimension=self.dimension,
+            namespace=self.namespace,
+            batch_size=self.batch_size,
+            **self.index_creation_kwargs,
+        )
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "PineconeDocumentStore":
+        return default_from_dict(cls, data)
+
+    def count_documents(self) -> int:
+        """
+        Returns how many documents are present in the document store.
+        """
+        try:
+            count = self._index.describe_index_stats()["namespaces"][self.namespace]["vector_count"]
+        except KeyError:
+            count = 0
+        return count
+
+    def write_documents(self, documents: List[Document], policy: DuplicatePolicy = DuplicatePolicy.NONE) -> int:
+        """
+        Writes Documents to Pinecone.
+
+        :param documents: A list of Documents to write to the document store.
+        :param policy: The duplicate policy to use when writing documents.
+            PineconeDocumentStore only supports `DuplicatePolicy.OVERWRITE`.
+
+        :return: The number of documents written to the document store.
+        """
+        if len(documents) > 0:
+            if not isinstance(documents[0], Document):
```

Comment:
If you want to be defensive you should also check that `documents` is a list. Why not EAFP?
```python
try:
    if not isinstance(documents[0], Document): raise TypeError()
except TypeError, KeyError:
    raise ValueError(msg)
```

BTW, we sure a malformed input is `ValueError`? I would raise `TypeError` in this case...